main.py

```
@@ -133,380 +255,529 @@ def __init__(
-                            """,
+        try:
+            # Construct messages for revision: Use base history, exclude intermediate AI calls, add proposal + critique
+            base_messages = state['messages']
+            # Filter out AI messages with tool calls that haven't been responded to
+            revision_base_messages = []
+            for i, msg in enumerate(base_messages):
+                # Add message if it's not an AI message with tool calls OR 
+                # if it IS an AI message with tool calls but the *next* message is a ToolMessage
+                if not (isinstance(msg, AIMessage) and msg.tool_calls):
+                     revision_base_messages.append(msg)
+                elif i + 1 < len(base_messages) and isinstance(base_messages[i+1], ToolMessage):
```

Ruff (0.8.2)

Undefined name `ToolMessage`

(F821)